### DISCUSSION TITLE
Issue with undefined symbol during kernel insertion pass (maybe :))

### DISCUSSION THREAD

**Discussion by jeromefereyre** [pocl/pocl#1558] [Q&A] [answered]
Hello,

I'm facing some issues with undefined symbols raised by pocl when modifying the embedded application to insert kernel function (if i have well understood).

My new device library, by the mean of intermediate calls, call the pocl function isLocalMemFunctionArg and cause this undefined symbol error.

	...
	[2024-09-10 08:57:44.206233429] PoCL: in fn int pocl_llvm_build_program(cl_program, unsigned int, cl_uint, _cl_program* const*, const char**, int) at line 760:
	 |      LLVM |  Writing program.bc to /home/user/.cache/pocl/uncached/_UNCACHED_EtIDSC/program.bc.
	[2024-09-10 08:57:44.268270638] PoCL: in fn int pocl_llvm_run_pocl_passes(llvm::Module*, _cl_command_run*, llvm::LLVMContext*, PoclLLVMContextData*, cl_kernel, cl_device_id, int) at line 1225:
	 |    TIMING |       >>>        21.321  ms    API: llvm_workgroup_ir_func_gen
	./opencl_test: symbol lookup error: /home/user/vrp_soft_dev_env/VXP_OpenCL/install/pocl_riscv/lib/pocl/libpocl-devices-XXX.so: undefined symbol: _ZN4pocl21isLocalMemFunctionArgEPN4llvm8FunctionEj
	make: *** [Makefile:58: runopencl] Error 127

But, this function seems to be available in the libOpenCL library i have compiled.

	$> nm /home/user/XXX/install/pocl_riscv/lib/libOpenCL.so.2.13.0 | grep isLocalMemFunctionArg
	0000000000100930 t _ZN4pocl21isLocalMemFunctionArgEPN4llvm8FunctionEj

And I call the opencl_test by setting LD_LIBRARY_PATH to find this libOpenCL.

I have tried in a new shell to set the LD_LIBRARY_PATH and display shared object dependencies 

	$> export LD_LIBRARY_PATH=/home/user/XXX/install/pocl_riscv/lib:/home/user/XXX/install/pocl_riscv/lib/pocl:/home/user/XXX/install/llvm/lib
	$> ldd /home/user/XXX/install/pocl_riscv/lib/libOpenCL.so.2.13.0
        libOpenCL.so.2 => /home/user/XXX/install/pocl_riscv/lib/libOpenCL.so.2 (0x00007f3c3fd28000)

and libOpenCL is resolved (libOpenCL.so.2 is linked to the right .so )

	ls -l /home/user/XXX/install/pocl_riscv/lib/libOpenCL.so.2
	lrwxrwxrwx 1 jf264787 360.1.361-M 19 Sep 10 10:55 /home/user/XXX/install/pocl_riscv/lib/libOpenCL.so.2 -> libOpenCL.so.2.13.0

The undefined symbol message let me think that the OpenCL is not found at runtime, since i have no compilation error.

But does anybody know why the problem is not solved by setting the LD_LIBRARY_PATH ? or what is the cause of this issue ? 

Thank you for reading.
Jerome

**Comment by pjaaskel** [comment, marked as answer]
Did you add POCL_EXPORT in LLVMUtils.h before its declaration (see the function above it for an example)?

**Reply by pjaaskel** [reply]
P.S. All RISC-V improving contributions back to the project are highly appreciated.

**Reply by jeromefereyre** [reply]
Great!  thank you for you answer, that was the solution ! :)
